README.txt
Introduction
============

This package is a Plone product that will add a `Page Turner` view to the PDF File content type. 
The result is that you'll be able to view PDFs much in the same way you can view them on
scribd.com. It uses the open source project, Flex Paper, to display the PDFs. You can find
information about it at `http://flexpaper.devaldi.com/`.


Requirements
------------

This product requires that you have pdf2swf installed. pdf2swf is included in older versions of SWFTools, 
up to 0.9.0 (the most recent version of SWFTools is 0.9.2 dating from 2012).


Installing pdf2swf with package managers
----------------------------------------

Where possible, we recommend you use package managers to install SWFTools as there are some dependencies
that need to be installed. 

If you're using Ubuntu Lucid (10.04) (and possibly later versions, but not Trusty Tahr, 14.04):

  # sudo apt-get install swftools

On Mac, if you have MacPorts installed you can,

  # sudo port install swftools

On Debian, you'll need to install from source as the swftools package does not include pdf2swf: `http://wiki.swftools.org/wiki/Installation`

Checking that you have installed pdf2swf
----------------------------------------

Once you have installed the package, check if pdf2swf is in your path. On Linux:

  # which pdf2swf
  
On Windows: make sure pdf2swf.exe is in your path:

  C:\> pdf2swf

If it isn't in your path (ie. you get an error message above), you'll have to install manually; see below.

Installing pdf2swf manually
---------------------------

If your package manager installs a version of SWFTools that does not include pdf2swf, you can install an older version of the package that does.

These instructions for installing on Ubuntu are based on http://serverfault.com/questions/623604/install-pdf2swf-on-ubuntu-trusty-tahr14-04.

For AMD64:

  # wget -P /tmp/ http://archive.canonical.com/ubuntu/pool/partner/s/swftools/swftools_0.9.0-0ubuntu2_amd64.deb
  # chmod a+x /tmp/swftools_0.9.0-0ubuntu2_amd64.deb
  # sudo dpkg -i /tmp/swftools_0.9.0-0ubuntu2_amd64.deb

For i386:

  # wget -P /tmp/ http://archive.canonical.com/ubuntu/pool/partner/s/swftools/swftools_0.9.0-0ubuntu2_i386.deb
  # chmod a+x /tmp/swftools_0.9.0-0ubuntu2_i386.deb
  # sudo dpkg -i /tmp/swftools_0.9.0-0ubuntu2_i386.deb
  
Other architectures are available at http://archive.canonical.com/ubuntu/pool/partner/s/swftools/

If all else fails, you can download SWFTools from `http://www.swftools.org/` and install it yourself.

How-To
------

* Add your PDF as a File to your Plone site (Add New -> File menu).

* Edit the various setting of the Page turner by clicking the `Page Turner Settings` button.

* To turn off auto-selecting of the page turner layout for PDF files, go to ZMI -> portal_properties -> site_properties and
  customize the page_turner_auto_select_layout property to off.
  
* PDFs that were on your site before you activated this add-on will need to be converted (you will see an
  error message on each File until you convert it). To convert an individual PDF, click the Flexpaper Convert 
  button. Large PDFs will take some time to be converted.  To convert all PDFs on your site, go to your site's 
  URL and append /@@convertall-to-flexpaper, e.g. http://www.yoursite.com/@@convertall-to-flexpaper


Tested With
-----------

* Plone 3 and 4

* Also works with Blob storage transparently so the converted PDFs aren't stored in the ZODB


Credits
-------

Credit goes to Wildcard Corp and Talin Senner for sponsoring and designing the product, and Nathan Van Gheem for coding it.


Asynchronous Conversion
-----------------------

With large PDFs the conversion to Flex Paper can take some time. If you have plone.app.async installed and configured, 
this conversion will happen asynchronously.


Convert All
-----------

If you'd like to convert all the existing documents on your site, visit the URL http://www.yoursite.com/@@convertall-to-flexpaper


TODO
----

- Add support for Dexterity content types. Maybe via a behavior.


Versions
--------

Flexpaper: 1.5.1


Upgrade Path
------------

For a more fully featured add-on, see https://pypi.python.org/pypi/collective.documentviewer 


docs/HISTORY.txt
Changelog
=========

1.3.8 ~ unreleased
------------------

- nothing yet

1.3.7 ~ 2015-06-03
------------------

- more detail for instructions
  [tkimnguyen]

- only queue conversion job if layout is page-turner
  [vangheem]


1.3.6 ~ 2013-05-07
------------------

- 4.3 template compatibility fix.
  [vangheem]


1.3.5 ~ 2013-05-07
------------------

- Remove zope.app.* dependencies.
  [esteele]

1.3.4 ~ 2012-10-02
------------------

- add back 1.3.2 change that was missed
  [vangheem]

1.3.3 ~ 2012-10-02
------------------

- changed 'cachebusting' to be false for all browsers
  [Joel.Kleier]

1.3.2 ~ 2012-07-31
------------------

- fix wmode used

1.3.1 ~ 2012-06-19
------------------

- fix uninstall profile


1.3.0 ~ 2012-04-11
------------------

- upgrade to flexpaper 1.5.1

1.2.5 ~ 2012-04-06
------------------

- fix js error when rendering pdf.
  [vangheem]

1.2.4 ~ 2011-10-16
------------------

- Plone 4.1 fix
  [vangheem]


1.2.3 ~ 2011-07-28
------------------

- fix default value for fit_width so the global setting
  will work now.
  [vangheem]
  

1.2.2 ~ 2011-04-01
------------------

- reference some of the resources off the portal root
  [vangheem]

1.2.1rc2 ~ 2011-04-01
---------------------

- fixes for windows users. 
  [fhurter]


1.2.1rc1 ~ 2011-03-14
---------------------

- fit page width for flex paper setting integration
  [espen]
  

1.2b1 ~ 2011-02-20
------------------

- re-convert pdf is password is provided on save of page turner settings


1.2a1 ~ 2011-02-15
------------------

- provide global settings configuration

- provide ability to enter password for encrypted pdf

- provide ability to convert all pdfs on the site to flex paper


1.1b1 ~ 2010-12-31
------------------

- work seamlessly with new version of wildcard.pdfpal so it will only
  create a flex paper after the new searchable PDF is provided.
  
- provide ability to provide extra command line options when converting the PDF.


1.0b4 ~ 2010-12-06
------------------

- add wmode transparent for flash params so that flash does not overlap
  other elements.
  
- switch to more accurate time format

- remove temp files even when there was an error


1.0b3 ~ 2010-12-02
------------------

- remove tmp files after using them on flex paper gen


1.0b2 ~ 2010-12-02
------------------

- added support for plone.app.async. Flex Paper conversion can now 
  happen in the background.
  [vangheem]
  

1.0b1 ~ 2010-11-29
------------------

- update flexpaper to 1.4.0
  [vangheem]
  
- auto select page turner view for pdf files now and make configurable
  via portal_properties -> site_properties
  [vangheem]
  
- better error checking when converting PDFs. pdf2swf still seems very
  touchy for converting some PDFs...
  [vangheem]
  
- Added progressive loading, print enabled, full screen visible, 
  search tools visible and cursor tools visible options
  [vangheem]


0.6
---

- Fix for storing blobs.
  [davisagli]

- Don't swallow exceptions while converting; log them.
  [davisagli]

- Use the subprocess module instead of popen2 to communicate with pdf2swf.
  [davisagli]

- Protect the page turner settings page with the "Modify view template"
  permission.
  [davisagli]

- Avoid purging existing view methods for the File type on installation.
  [davisagli]

0.5 ~ 2010-05-21
----------------

- Initial release


wc/pageturner/views.py
from Products.Five.browser import BrowserView
from plone.app.form import base as ploneformbase
from zope.interface import implements
from zope.formlib import form
from webdav.common import rfc1123_date
from Products.Archetypes.utils import contentDispositionHeader
from os import fstat
from Products.CMFCore.utils import getToolByName
from interfaces import IPageTurnerSettings, IUtils, IGlobalPageTurnerSettings
from wc.pageturner import mf as _
from settings import Settings, GlobalSettings
from plone.app.contenttypes.interfaces import IFile
import zope.event
import zope.lifecycleevent
from zope.component import getMultiAdapter
from convert import pdf2swf
from zope.component.hooks import getSite
from plone.memoize.view import memoize
from wc.pageturner.events import queue_job
from DateTime import DateTime

from logging import getLogger
logger = getLogger('wc.pageturner')

from plone.app.blob.download import handleRequestRange
from plone.app.blob.iterators import BlobStreamIterator
from plone.app.blob.utils import openBlob


def either(one, two):
    if one is None:
        return two
    return one


class PageTurnerView(BrowserView):

    installed = pdf2swf is not None
    enabled = pdf2swf is not None

    @property
    @memoize
    def portal_url(self):
        return getMultiAdapter((self.context, self.request),
                               name="plone_portal_state").portal_url()

    def initialize(self):
        site = getSite()
        self.settings = Settings(self.context)
        self.global_settings = GlobalSettings(site)

        utils = getToolByName(self.context, 'plone_utils')
        msg = None

        if self.context.file.contentType in ('application/pdf',
                                             'application/x-pdf', 'image/pdf'):
            if not self.installed:
                msg = "Since you do not have swftools installed on this " + \
                      "system, we can not render the pages of this PDF."
            elif self.settings.converting is not None and \
                    self.settings.converting:
                msg = "The PDF is currently being converted to the " + \
                      "FlexPaper view..."
                self.enabled = False
            elif not self.settings.successfully_converted:
                msg = "There was an error trying to convert the PDF. Maybe " +\
                      "the PDF is encrypted, corrupt or malformed?"
                self.enabled = False
        else:
            self.enabled = False
            msg = "The file is not a PDF. No need for this view."

        if msg:
            mtool = getToolByName(self.context, 'portal_membership')
            if mtool.checkPermission('cmf.ModifyPortalContent', self.context):
                utils.addPortalMessage(msg)

    def __call__(self):
        self.initialize()
        return super(PageTurnerView, self).__call__()

    def javascript(self):
        return """
jQuery(document).ready(function(){
    window.flexPaper = new FlexPaperViewer(
        '%(portal_url)s/%(resources)s/FlexPaperViewer',
        'pageturner', { config : {
        SwfFile : escape('%(context_url)s/converted.swf'),
        Scale : 0.6,
        ZoomTransition : 'easeOut',
        ZoomTime : 0.5,
        ZoomInterval : 0.2,
        FitPageOnLoad : true,
        FitWidthOnLoad : %(fit_width_on_load)s,
        FullScreenAsMaxWindow : false,
        ProgressiveLoading : false,
        MinZoomSize : 0.2,
        MaxZoomSize : 5,
        SearchMatchAll : false,
        InitViewMode : 'Portrait',
        PrintPaperAsBitmap : false,

        ViewModeToolsVisible : true,
        PrintEnabled : %(print_enabled)s,
        CursorToolsVisible : %(cursor_tools_visible)s,
        SearchToolsVisible : %(search_tools_visible)s,
        ProgressiveLoading : %(progressive_loading)s,

        localeChain: 'en_US',
        swfinstall: '%(portal_url)s/%(resources)s/playerProductInstall.swf'
        }});
    jQuery('#pageturner').show().width(%(width)s).height(%(height)s);
});
""" % {
            'resources': '++resource++pageturner.resources',
            'context_url': self.context.absolute_url(),
            'portal_url': self.portal_url,
            'width': either(self.settings.width, self.global_settings.width),
            'height': either(self.settings.height, self.global_settings.height),  # noqa
            'fit_width_on_load': str(either(self.settings.fit_width_on_load,
                self.global_settings.fit_width_on_load)).lower(),
            'progressive_loading': str(either(self.settings.progressive_loading,  # noqa
                self.global_settings.progressive_loading)).lower(),
            'print_enabled': str(either(self.settings.print_enabled,
                self.global_settings.print_enabled)).lower(),
            'full_screen_visible': str(either(self.settings.full_screen_visible,  # noqa
                self.global_settings.full_screen_visible)).lower(),
            'search_tools_visible': str(either(self.settings.search_tools_visible,  # noqa
                self.global_settings.search_tools_visible)).lower(),
            'cursor_tools_visible': str(either(self.settings.cursor_tools_visible,  # noqa
                self.global_settings.cursor_tools_visible)).lower(),
}


class DownloadSWFView(PageTurnerView):

    def __call__(self):
        self.initialize()
        # done much like it is done in plone.app.blob's index_html
        header_value = contentDispositionHeader(
            disposition='inline',
            filename=self.context.file.filename.replace('.pdf', '.swf'))

        blob = self.settings.data
        blobfi = openBlob(blob)
        length = fstat(blobfi.fileno()).st_size
        blobfi.close()

        self.request.response.setHeader(
            'Last-Modified', rfc1123_date(self.context._p_mtime))
        self.request.response.setHeader('Accept-Ranges', 'bytes')
        self.request.response.setHeader('Content-Disposition', header_value)
        self.request.response.setHeader("Content-Length", length)
        self.request.response.setHeader(
            'Content-Type', 'application/x-shockwave-flash')
        range = handleRequestRange(
            self.context, length, self.request,
            self.request.response)
        return BlobStreamIterator(blob, **range)


class SettingsForm(ploneformbase.EditForm):
    """
    The page that holds all the slider settings
    """
    form_fields = form.FormFields(IPageTurnerSettings)

    label = _(u'heading_pageturner_settings_form',
              default=u"Page Turner Settings")
    description = _(u'description_pageturner_settings_form',
                    default=u"these settings override the global settings.")

    @form.action(_(u"label_save", default="Save"),
                 condition=form.haveInputWidgets,
                 name=u'save')
    def handle_save_action(self, action, data):
        if form.applyChanges(self.context, self.form_fields, data,
                             self.adapters):
            zope.event.notify(
                zope.lifecycleevent.ObjectModifiedEvent(self.context))
            zope.event.notify(ploneformbase.EditSavedEvent(self.context))
            self.status = "Changes saved"
        else:
            zope.event.notify(ploneformbase.EditCancelledEvent(self.context))
            self.status = "No changes"

        # convert right now if password provided
        if data.get('encryption_password', None):
            settings = Settings(self.context)
            settings.last_updated = DateTime('1999/01/01').ISO8601()
            queue_job(self.context)

        url = getMultiAdapter((self.context, self.request),
                              name='absolute_url')() + '/view'
        self.request.response.redirect(url)


class GlobalSettingsForm(ploneformbase.EditForm):
    form_fields = form.FormFields(IGlobalPageTurnerSettings)

    label = _(u'heading_pageturner_global_settings_form',
              default=u"Global Page Turner Settings")
    description = _(u'description_pageturner_global_settings_form',
                    default=u"Configure the parameters for this Viewer.")

    @form.action(_(u"label_save", default="Save"),
                 condition=form.haveInputWidgets,
                 name=u'save')
    def _handle_save_action(self, action, data):
        if form.applyChanges(self.context, self.form_fields, data,
                             self.adapters):
            zope.event.notify(ploneformbase.EditSavedEvent(self.context))
            self.status = "Changes saved"
        else:
            zope.event.notify(ploneformbase.EditCancelledEvent(self.context))
            self.status = "No changes"
        self.request.response.redirect(
            self.context.absolute_url() + '/@@global-page-turner-settings')


class Utils(BrowserView):
    implements(IUtils)

    def enabled(self):
        try:
            return IFile.providedBy(self.context) and \
                self.context.getLayout() == 'page-turner'
        except:
            return False

    def convert(self):
        if self.enabled():
            settings = Settings(self.context)
            settings.last_updated = DateTime('1999/01/01').ISO8601()
            queue_job(self.context)

        self.request.response.redirect(self.context.absolute_url() + '/view')

    def convert_all(self):
        confirm = self.request.get('confirm', 'no')
        if confirm != 'yes':
            return 'You must append "?confirm=yes"'
        else:
            ptool = getToolByName(object, 'portal_properties')
            site_props = getattr(ptool, 'site_properties', None)
            auto_layout = site_props.getProperty(
                'page_turner_auto_select_layout', False)

            catalog = getToolByName(self.context, 'portal_catalog')
            files = catalog(object_provides=IFile.__identifier__)
            for brain in files:
                file = brain.getObject()
                if file.file.contentType not in ('application/pdf',
                                                 'application/x-pdf',
                                                 'image/pdf'):
                    continue

                if auto_layout and file.getLayout() != 'page-turner':
                    file.setLayout('page-turner')

                self.request.response.write(
                    'Converting %s to flex paper...\n' % file.absolute_url())
                settings = Settings(file)
                settings.last_updated = DateTime('1999/01/01').ISO8601()
                queue_job(file)
